Updates for Dictionaries Precourse lesson (#110)
Ada-Developers-Academy/ada-precourse-v2: intro-to-python/dictionaries.md

Reviewer: Line 565:
```
The logical error comes in because all of the letters from the word are in the word_dict! If we only check to see if the letter is in the word_dict, we'll tell the user that they've already guessed a letter even if it's the first time they've guessed it if it's in the word! We need to add a check that looks at the value to see if it's `True` or `False`:
```
There are some places we can use "whether" and there's an awkward sentence (maybe `snowman_dict`?)
```
The logical error comes in because all of the letters from the word are already in `word_dict`, just with a `False` value! If we only check to see whether the letter is in `word_dict`, we'll tell the user that they've already guessed a letter even if it's the first time they've guessed it! We need to add a check that looks at the value to see whether it's `True` or `False`:
```

Author: I updated the sentence to use "whether" but I don't get what you mean about the awkward sentence. Do you mean the variable `word_dict` should be `snowman_dict`?

Reviewer: Sorry, it's hard to see the "awkward" bit wihtout the suggestion highlighting. There was a part that went
> even if it's the first time they've guessed it if it's in the word!
and I changed it to
> even if it's the first time they've guessed it!

The `snowman_dict` was a potential separate suggestion, but that should be made in consideration of whether it's overall more or less confusing to refer to the `word_dict` variable before we've talked about where it appears in the final updated code. Maybe we could add a little text earlier that makes it more clear that `word_dict` will be the updated parameter name or something like that. But I'm fine with leaving that, though I feel like some of the naming in our sample solution could use a revisit overall. (why was it ever OK that there was a `list1` and `list2`?)

Reviewer: Line 738:
```
1. The game prints out a section of the snowman drawing, showing one level for each incorrectly guessed letters
```
has a typo (letters/letter)
```
1. The game prints out a section of the snowman drawing, showing one level for each incorrectly guessed letter
```

Reviewer: line 447:
```
Two working implementations:
```
should be 
```
An example solution:
```
or something to that effect (singular)

### Dictionaries, A New Kind of Data Structure

In the lists lesson we used our first data structure.  Lists are powerful and useful tools, but they do have some limitations.  Say we want to use our list to store all of the pieces of an address.  We could use several variables to store the city, state, zip code, street address, but as we saw in the list lesson, being able to store all of the information in one variable can make it simpler to access the data and pass the data to functions.  We can use a list to store each of these pieces of information, but we would need to remember the location where we stored them so we could access the elements by their index.  

> We could set up constants like:

Author: Yeah I think this point would be best covered in another ticket for us to pick up at another time. Thank you!

I agree though. Before looking over this lesson for this ticket, I didn't know we covered homogenous vs heterogenous dictionaries and as the lesson is written it kind of feels like it's brought up just for the sake of mentioning it.

> We could set up constants like:

Reviewer: Asana ticket: https://app.asana.com/0/1201700438643002/1208039033382219/f

## Vocabulary and Syntax

| Vocab          | Definition                                                    | Synonyms  | How to Use in a Sentence                                                      |
| -------------- | ------------------------------------------------------------- | --------- | ----------------------------------------------------------------------------- |
| dictionary | A collection that associates unique keys with values | hash table, associative array | "This dictionary uses a planet name as the key and the value is its distance from the sun." |
| key | A key is used to retrieve a value stored in a dictionary. Keys in a dictionary must be unique. | address | "The keys for the dictionary are all the planets, 'mercury', 'venus', 'mars', and so on." |
| value | A piece of data of any possible type. Values in a dictionary, unlike keys, are not necessarily unique. | content | "The value for the key 'mercury' is '36 million miles'."|

> Let's look at some Python syntax related to dictionaries. The snippet below shows how we can create an empty dictionary, create a dictionary with key-value pairs in it, add a key-value pair to a dictionary, and how to access a value from a dictionary with a key.

Author: I'm thinking something like: 

```
### !callout-info

## Adding New Values to a Dictionary

Notice that we do not use a method to add a value to a dictionary in the example above. Unlike lists, which have an `append` method for adding new elements, dictionaries do not have a method for adding key-value pairs. Instead, dictionaries use square bracket notation to declare a new key and the assignment operator to assign a value. 

### !end-callout
```

How's that sound?

<!--optional-->
##### !explanation

Two working implementations:
An example solution:

```python

def get_value_from_dictionary(dict, key):
    if key in dict:
        return dict[key]
    else:
        return None

def get_value_from_dictionary(dict, key):
    try:
        return dict[key]
    except KeyError:

>         return None

Author: We never go into detail about what try/except is in Precourse. It feels funny to throw this example in here without explaining the concept. 

Do you think we should remove it or leave a comment introducing it and let students explore their imagination? It does come up later in Unit 1.

Also, how do you feel about `return None` in the `else`? In the change I removed else, but can revert the change if you prefer it nested in there.

```

We can use the `not` operator with the `in` operator too. 
```py
# True or False depending on whether some_key does not exist as a key in the dictionary some_dict
is_some_key_in_some_dict = some_key not in some_dict
```

In general, the `in` operator checks whether some data is included in a data structure or string. For dictionaries, `in` checks whether a key is in a dictionary. If the key is present then the expression evaluates to `True` and if the key is absent then the expression evaluates to `False`.

##### !end-hint

<!--optional-->
##### !explanation

Some example solutions:

```python

def get_value_from_dictionary(dict, key):
    if key not in dict:
        return None
	        
    return dict[key]
    
def get_value_from_dictionary(dict, key):
    if key in dict:
        return dict[key]

    return None
```

Both solutions work and are very similar! Pay close attention to the `if` statement and what is returned from each approach. We prefer the first solution because using the key to access a value from the dictionary and returning it is the main logic. This logic is **not** nested inside a conditional block and is less indented. 

> 

Reviewer: Just needs a break
```suggestion

<br>

```

1. If the dictionary contains the key, return the value associated with that key
1. If the dictionary does not contain the key, return None

<br/>

To check whether a key is in a dictionary, we can use the `in` operator.

```py
# True or False depending on whether some_key exists as a key in the dictionary some_dict
is_some_key_in_some_dict = some_key in some_dict
```

In general, the `in` operator checks whether some data is included in a data structure or string. For dictionaries, `in` checks whether a key is in a dictionary. If the key is present then the expression evaluates to `True` and if the key is absent then the expression evaluates to `False`.

##### !end-hint

<!--optional-->
##### !explanation

An example solution:

```python

> def get_value_from_dictionary(dict, key):
>     if key in dict:
>         return dict[key]
> 
>     return None

Reviewer: The suggestion ends up looking weird, but one more thought here is to fully lean in to the guard clause feel here, making the key access the main logic flow (less indented) over the None return.

I'm sure we'd prefer not messing around with this example any more, but I think I can say with confidence, this is how I would encourage folks to write it. However, this also requires rephrasing the hint, and possibly bringing back a second example if we'd want to contrast the affirmative and negative condition approach.

What do you think?

```suggestion
def get_value_from_dictionary(dict, key):
    if key not in dict:
        return None
        
    return dict[key]
```

Author: How about re-writing the hint like this:

```
##### !hint

A solution for this problem can be broken into the following steps:

1. Check to see if the dictionary does **not** contain the key
1. If the dictionary does not contain the key, return None
2. If the dictionary does contain the key, return the value associated with that key

<br/>

To check whether a key is in a dictionary, we can use the `in` operator.

```py
# True or False depending on whether some_key exists as a key in the dictionary some_dict
is_some_key_in_some_dict = some_key in some_dict
```

We can use the `not` operator with the `in` operator too. 
```py
# True or False depending on whether some_key does not exist as a key in the dictionary some_dict
is_some_key_in_some_dict = some_key not in some_dict
```

In general, the `in` operator checks whether some data is included in a data structure or string. For dictionaries, `in` checks whether a key is in a dictionary. If the key is present then the expression evaluates to `True` and if the key is absent then the expression evaluates to `False`.

##### !end-hint
```

And then re-writing the examples like this:
```
##### !explanation

Some example solutions:

```python

def get_value_from_dictionary(dict, key):
    if key not in dict:
        return None
	        
    return dict[key]
    
def get_value_from_dictionary(dict, key):
    if key in dict:
        return dict[key]

    return None
```

Both solutions work and are very similar! Pay close attention to the `if` statement and what is returned from each approach. We prefer the first solution because using the key to access a value from the dictionary and returning it is the main logic. This logic is **not** nested inside a conditional block and is less indented. 

While the second example works, it nests the logic of getting a value inside a conditional block and `return None` is emphasized as the main logic. Writing the code this way does not quite capture our intent to return some value from a dictionary.

##### !end-explanation
```

Reviewer: That looks great! It sounds a little samey to the later one but I think that’s ok since this is a hint, so we can’t be sure that everyone will read them. Having each stand on its own feels like the right thing to do.

    return dict

def dict_counter(dict, key):
    try:
        dict[key] += 1
    except KeyError:
        dict[key] = 1

    return dict

> 

Author: Same question as above ^^

        for index in range(len(keys)):
            result[keys[index]] = values[index]
        return result

>     return None

Author: I removed the second approach.

>     return None

Author: Thanks, with the duplicate code cleaned up and with zip out of the picture, we're left with: 
```
def build_a_dictionary(keys, values):
    if len(keys) != len(values):
        return None
      
    result = {}
    for index in range(len(keys)):
        result[keys[index]] = values[index]
    return result
```

</details>

### Using the Word Dictionary

Now that we have the `build_word_dict` function we need to call it in our `snowman()` function and replace the `correct_guesses_list` with the new dictionary that's generated by `build_word_dict`. 

We will need to:
1. Remove the `correct_guesses_list` since we won't be using it anymore
2. Invoke `build_word_dict` and pass in `snowman_word` as an argument. We should use a variable called `snowman_dict` to capture the return value from calling the function.
3. Refactor the call to `get_letter_from_user` so that the first argument is `snowman_dict` instead of `correct_guesses_list`, but we'll still need `wrong_guesses_list` so we can leave the second argument as is.
4. Now that we have `snowman_dict`, we should use it to see if a user has correctly guessed a letter that is in the secret `snowman_word`  
5. If the user's guess is correct, then we need to update `snowman_dict` dictionary. Before a correct guess, a key (represented by a letter) has a value set to `False`. After a correct guess, we need to update the key's value to `True`

>    1. Since we're using `snowman_dict` instead of `correct_guesses_list`, we can now remove the line of code `correct_guesses_list.append(user_input)` 

Author: I added more details because I felt like the change is a little complicated. What do you think?

    return word_dict
```

The end result here will be the same, in the first example any time we encounter a letter that is already in `word_dict` we'll overwrite it, but the values are always the same, so the end result of both of functions above will be the same.
</details>

### Using the Word Dictionary

Now that we have the `build_word_dict` function we need to call it in our `snowman()` function and replace the `correct_guesses_list` with the new dictionary that's generated by `build_word_dict`. 

We will need to:
1. Remove the `correct_guesses_list` since we won't be using it anymore

> 2. Invoke `build_word_dict` and pass in `snowman_word` as an argument. We should use a variable called `snowman_dict` to capture the return value from calling the function.

Author: Could we open another card to make this change? I know it's a small one here, but we'd need to make sure it's applied too the example code we link off to and also in the replit. 

I did it for the elem to letter_num suggestion, but I'm afraid of mucking it up even more and introducing some inconsistencies. 

I'd like to hold off on this one for now, get these changes merged in, check for continuity across all the lessons, and then make this change.

> 2. Invoke `build_word_dict` and pass in `snowman_word` as an argument. We should use a variable called `snowman_dict` to capture the return value from calling the function.

Reviewer: Asana task: https://app.asana.com/0/1201700438643002/1208039036671787/f

The next step is to update our helper function `get_letter_from_user` to use the word dictionary instead of the list of correct letters.  Right now, we're only telling a user if they've guessed a letter before, but while we're updating this function let's modify it to tell the user if they've guessed a letter before, __and__ if that letter is in the word or not.  

There is one important change to be aware of, right now we're just checking if the letter the user guessed is in the list of correct letters.  If we do that with the dictionary, we'll end up with a logical error!  Consider these code snippets:

```python

#snippet from the version of get_letter_from_user:
# Snippet from the previous version of get_letter_from_user:

#...

> elif user_input_string in list1 or user_input_string in list2:

Author: Somewhere along the line the variable names were changed to list1 and list2.... Changing them back to what we introduced them as in the Lists lesson.

</details>

The next step is to update our helper function `get_letter_from_user` to use the word dictionary instead of the list of correct letters.  Right now, we're only telling a user if they've guessed a letter before, but while we're updating this function let's modify it to tell the user if they've guessed a letter before, __and__ if that letter is in the word or not.  

> There is one important change to be aware of, right now we're just checking if the letter the user guessed is in the list of correct letters.  If we do that with the dictionary, we'll end up with a logical error!  Consider these code snippets:

Reviewer: Also, a following paragraph that wasn't part of this change could use some fo the same changes (added as a file-level comment).

```python

# Snippet from the previous version of get_letter_from_user:

#...
elif user_input_string in wrong_guesses_list or user_input_string in correct_guesses_list:

>     print("You already guessed that letter")

Reviewer: mirrors a change a little further down
```suggestion
    print("You have already guessed that letter")
```

#...

# assume we replace the parameter `correct_guesses_list` in the function call with the parameter `word_dict`
# if we just swap it in we'll end up with a logical error:

#...
elif user_input_string in wrong_guesses_list or user_input_string in word_dict:

>     print("You have already guessed that letter")

Reviewer: Since `get_letter_from_user`hasn't been changed yet (so the param name hasn't changed) should we stick with `snowman_dict` rather than `word_dict` for this thought experiment?

I could also imagine `snowman_dict` could be confusing when they actually update the function, so keeping it as is is fine too.

Author: This suggestion can be addressed by: https://app.asana.com/0/1201700438643002/1208039036671787/f

```python

#...

> elif user_input_string in wrong_guesses_list or (user_input_string in word_dict and word_dict[user_input_string]):

Reviewer: potentially `snowman_dict` rather than `word_dict`.

But gosh this looks complicated without using `.get()` *(no change requested)*

#...
elif user_input_string in wrong_guesses_list or user_input_string in correct_guesses_list:
    print("You have already guessed that letter")
#...

# assume we replace the parameter `correct_guesses_list` in the function call with the parameter `word_dict`
# if we just swap it in we'll end up with a logical error:

#...
elif user_input_string in wrong_guesses_list or user_input_string in word_dict:
    print("You have already guessed that letter")
#...

```
The logical error comes in because all of the letters from the word are already in `word_dict`, just with a `False` value! If we only check to see whether the letter is in `word_dict`, we'll tell the user that they've already guessed a letter even if it's the first time they've guessed it! We need to add a check that looks at the value to see whether it's `True` or `False`:

```python

#...
elif user_input_string in wrong_guesses_list or (user_input_string in word_dict and word_dict[user_input_string]):
    print("You have already guessed that letter")
#...

```

Update `get_letter_from_user` helper function and make the following changes:
1. Change the first parameter to be the word dictionary instead of a list

> 2. Add additional feedback to the user when they attempt to guess a letter that they have already guessed that informs them if the letter is in the word or not.

Reviewer: avoids two that clauses (plus if/whether)
```suggestion
2. Add additional feedback to the user, when they attempt to guess a letter that they have already guessed, to inform them whether or not the letter is in the word.
```

### Displaying User Progress

Now we're going to get back to our original goal, displaying each letter of the word with an underscore `_` if the letter has not yet been guessed and the correct letter if it has been guessed.  For example, if the word is `pepper` and our user has guessed the letters `p`, and `r`, we want to print out `p _ p p _ r`.  Each time they guess a correct letter, we want to print out an updated version of this string.  We're going to start by writing a helper function that generates and returns this string.  This function will need to:

* Create an empty string
* Loop over each letter in the word

> * Check if the letter has been guessed or not

Reviewer: possible whether
```suggestion
* Check whether the letter has been guessed or not
```

    * If the value of "guessed" is `True` add the letter to the string

>     * If the value of "guessed" is `False` add `_` to the string

Reviewer: I'm sure we're avoiding a list and joining for simplicity, but this really hurts. 😭

* Add 1 space between each letter/underscore

In `snowman.py` write a helper function `generate_word_progress_string` that takes two parameters, a word and a dictionary where each letter in the word is a key and the values are `True` or `False`.  The function returns a string that represents the word, with spaces between each letter.  For each letter, if the value in the dictionary is `True`, the letter is displayed.  If the value in the dictionary is `False`, the letter is replaced with a `_` character.

Example inputs and outputs:

|input|output|
|--|--|
|`word="pepper"` <br/> `word_dict={"p":True, "e": False, "r": False}`|`"p _ p p _ _"`|
|`word="tiger"` <br/> `word_dict={"e":False, "g": False, "i": False, "r": False, "t": False}`|`"_ _ _ _ _"`|
|`word="swamp"` <br/> `word_dict={"a":True, "m": True, "p": True, "s": True, "w": True}`|`"s w a m p"`|

### !callout-info

> ## Write Some Code, Check the Code, Repeat

Reviewer: We need an explicit `<br />` between each pair of paragraphs so they don't get crushed together in the output.

> As you're writing your implementation, be sure to call the helper function often so that you can see the output and confirm that it's working or not working. It's good practice to write a line or logical unit of code and then run it to check that you are getting expected output before moving on to the next line of code. It is easier to identify where a bug is in one or two lines of code versus ten lines of code.

Reviewer: a "whether" for something that wasn't "if" 🙂  
```suggestion
As you're writing your implementation, be sure to call the helper function often so that you can see the output and confirm whether or not it's working. It's good practice to write a line or logical unit of code and then run it to check that you are getting the expected output before moving on to the next line of code. It is easier to identify where a bug is in one or two lines of code versus ten lines of code.
```

If you run into a bug, add debugging print statements or walk through the code line by line and make predictions about what should happen versus what actually happens to help you resolve the bug. 

Finally, recall that the helper function should __return__ a string. When you invoke the function, if you're not seeing any output printed to the terminal you will need to print the value returned from `generate_word_progress_string`.

As always, feel free to reach out over Slack if you need help!

### !end-callout

<br>

<details>

> <summary>Here is our implementation.</summary>

Reviewer: Would it be too big to swap `elem_num` and `elem` to `letter_num` and `letter`?

Author: Yeah, I can do that. Noting here that I'll also be making the change in the code example (https://github.com/Ada-Developers-Academy/ada-precourse-v2/blob/main/intro-to-python/resources/src/dictionaries/games_dictionary.md) and also in snowman.py in replit.

<br/>

If you run into a bug, add debugging print statements or walk through the code line by line and make predictions about what should happen versus what actually happens to help you resolve the bug. 

<br/>

Finally, recall that the helper function should __return__ a string. When you invoke the function, if you're not seeing any output printed to the terminal you will need to print the value returned from `generate_word_progress_string`.

<br/>

As always, feel free to reach out over Slack if you need help!

### !end-callout

<br>

<details>
<summary>Here is our implementation.</summary>

```python

def generate_word_progress_string(word, word_dict):
    output_string = ""
    letter_num = 0

    for letter in word:
        if letter_num > 0:
            output_string += " "

        if word_dict[letter]:
            output_string += letter
        else:
            output_string += "_"

        letter_num += 1

    return output_string
    
```

</details>

> Now that we have the helper function `generate_word_progress_string` working, let's revisit our original goal.  Right now we're returning the progress string, but what we really want to do is display it. Let's switch to printing out the result instead of returning it. We'll rename our function to `print_word_progress_string`:

Reviewer: Also need the enum → letter swap in the code excerpt for `print_word_progress_string` below here.

    
```

</details>

> Now that we have the helper function `generate_word_progress_string` working, let's revisit our original goal.  Right now we're returning the progress string, but what we really want to do is display it. Let's switch to printing out the result instead of returning it. We'll rename our function to `print_word_progress_string`:

Reviewer: Possibly add `print_word_progress_string` and have it *call*`generate_word_progress_string` rather than renaming and changing the behavior (separation of concerns)?

Author: This also has cascading changes: the code example, replit, snowman project directions (https://learn-2.galvanize.com/cohorts/2330/blocks/1151/content_files/intro-to-python/snowman.checkpoint.md?assessment=true). 

Hoping we can lump this change in with the name changes so that nothing falls through the cracks with this PR

Reviewer: Asana task: https://app.asana.com/0/1201700438643002/1208039036671787/f

Reviewer: Asana ticket: https://app.asana.com/0/1201700438643002/1208039036671787/f

    for elem in word:
       # ...
       
    # return output_string
    print(output_string)

```

### Get Game Progress

> The helper function `print_word_progress_string` provides a visual display of the players progress towards guessing the word. We also need to write a function that indicates whether or not all the letters of the word have been guessed. We will name this function `get_word_progress`. The function `get_word_progress` should return `True` if all the letters have been guesses, and `False` otherwise.

Reviewer: typos
```suggestion
The helper function `print_word_progress_string` provides a visual display of the player's progress towards guessing the word. We also need to write a function that indicates whether or not all the letters of the word have been guessed. We will name this function `get_word_progress`. The function `get_word_progress` should return `True` if all the letters have been guessed, and `False` otherwise.
```

Reviewer: (I would also love this function to be called `is_word_guessed`.)

Author: We can add this rename from `get_word_progress` to `is_word_guessed` to the work to rename `snowman_dict` to `guessed_dict`?

Reviewer: Asana task: https://app.asana.com/0/1201700438643002/1208039036671787/f
